### PR TITLE
Fix for issue 42

### PR DESCRIPTION
**Pull request by rpgoldman**
The setting of default values for *dot-path* and *neato-path* was breaking CLISP.

Fixes #47

### CODE REVIEW

**Comment by rpgoldman** [comment]
All interested persons, *please review and comment*!  There are no automated tests, so I have no way to validate my patch before merging.

**Comment by ko56** [comment]
> /cc @ko56 -- if you get a chance, please try this out and comment to let me know if it works for you.

Looks like it works.

**Comment by mdbergmann** [comment]
SBCL 2.4.1 and LispWorks 8.0.1 can find the path on macOS. So it's OK.
But I tried to rename 'dot' to force the `(warn ...`.  But I've not seen a warn in the repl. The `*dot-oath*` was `NIL` then.

**Comment by hraban** [comment]
Successfully built with ECL, SBCL & CLISP at https://github.com/hraban/cl-nix-lite/actions/runs/7921025641 . Just the build, since there are no tests, but still an improvement 👍  thanks!

**Comment by hraban** [comment]
PS the title of the PR has a typo; it fixes #42 not #47 :)

**Comment by rpgoldman** [comment]
> SBCL 2.4.1 and LispWorks 8.0.1 can find the path on macOS. So it's OK. But I tried to rename 'dot' to force the `(warn ...`. But I've not seen a warn in the repl. The `*dot-oath*` was `NIL` then.

This occurs because there used to be two _completely different_ code paths for locating "dot" and "neato."  I dropped one of them, but that may have removed the warning.  I will see about restoring it before merging.

**Comment by rpgoldman** [comment]
I just checked and the warn seems to work for me:

Allegro:
```
...
;    /Users/rpg/.cache/common-lisp/acl-11.0-macosx-arm64/Users/rpg/lisp/cl-dot/cl-dot.fasl
Warning: Could not find 'dot' executable! Please set CL-DOT:*DOT-PATH*
Warning: Could not find 'neato' executable! Please set
         CL-DOT:*NEATO-PATH*
...
```

SBCL:
```
; wrote /Users/rpg/.cache/common-lisp/sbcl-2.4.0-macosx-arm64/Users/rpg/lisp/cl-dot/cl-dot-tmpB2JTL1W9.fasl
; compilation finished in 0:00:00.035
WARNING: Could not find 'dot' executable! Please set CL-DOT:*DOT-PATH*
WARNING: Could not find 'neato' executable! Please set CL-DOT:*NEATO-PATH*
; compiling file "/Users/rpg/lisp/cl-dot/deprecated.lisp" (written 24 MAY 2016 11:29:07 AM):
```

I did this by doing:
```
PATH=<normal path with /opt/homebrew/bin, where my copies of dot and neato are located, removed> sbcl
```
then required ASDF, `(asdf:load-asd "cl-dot.asd")` and `(asdf:load-system "cl-dot")`.

**Comment by rpgoldman** [comment]
The warnings appear in the middle of the transcript from `load-system`, which is not ideal.  I *believe* we could move the setting of `*dot-path*` and `*neato-path*` (but *not* the `defvar`s) to the final stage of the ASDF system, so that the warnings would appear more visibly at the end of the load transcript.

I'd accept a tweak to this MR to make that change, or will try to do it myself, but won't get to that until next week sometime.

**Comment by rpgoldman** [comment]
One more concern: calls to `WARN` cause `load-system` to fail (because they cause `cl:compile-file` to return `nil`). That probably is *wrong* for cl-dot.  CL-DOT can quite happily be used to produce dot files but not actually run dot or neato.  In fact, I *never* run "dot" from inside CL: I always just write the ".dot" output file and then translate to an output format using the command line.

So:  would it be better to call `(format t "NOTE: ...")` instead of `warn` here?

**Comment by mdbergmann** [comment]
Hmm, yeah, maybe.
I have CL actually call 'dot' to generate the output. But anyway, a 'note' is fine.

**Comment by hraban** [comment]
Just for my reference: all this rigmarole is ultimately only really necessary for Windows, right? Every "normal" system would just be expected to have graphviz on its PATH and just use `(run-process '("dot" ...) :search t)`?

**Comment by Zulu-Inuoe** [comment]
@hraban No. Look at #42 - this is an issue on Mac

**Comment by hraban** [comment]
But I mean the _entire_ juggling of hard-coded paths, all of it, is only necessary because we expect some systems to not have a sane "PATH", right? What UNIX system would have dot installed, but not in PATH? Yes we're now seeing errors on Mac because of some faulty assumptions in the implementation, but at the end of the day isn't this all a work-around for a problem that only really occurs on Windows to begin with?

I'm just checking if this is true: if it weren't for windows, there would be no hard-coding of paths or reading of the PATH variable at all--this lib would just call `(run-process '("dot" ...) :search t)` and either get a result or no result. Is that correct?

**Comment by Zulu-Inuoe** [comment]
Given the original code was
```lisp
#+os-macosx
(eval-when (:compile-toplevel :load-toplevel :execute)
  (defvar +mac-binary-path+
    (or (probe-file "/opt/homebrew/bin/") ; homebrew
        (probe-file "/opt/local/bin/")  ; MacPorts
        (probe-file "/usr/bin/"))))     ; best guess
```
it seems that it's also a problem on Mac systems

It's anecdotal of course, but I'm on Windows and `dot` is certainly on my PATH

**Comment by hraban** [comment]
Ok I'm not being clear so let me rephrase my question:

Why are paths to `*/bin` directories hard-coded at all? Why is the `PATH` envvar read? Any other library dealing with calling third party tools would just use `execvp` aka `run-program ... :search T` aka let the system handle it.

**Comment by Zulu-Inuoe** [comment]
If I were the author of cl-dot, I agree, I would just pull from PATH, because where the executable lives is up to the user, not me. I think for convenience, I'd also allow cl-dot related functions to accept an alternative location for it as a parameter.

But that's up to the maintainers here, not me, and the code exists to search for it in some locations, and presumably that was driven by some need somewhere, sometime

**Comment by hraban** [comment]
Exactly.

The question is: what was that original reason?

It might be worthwhile to figure that out before gluing more wallpaper over it.

(I'm putting 3¢ on "windows".)

**Comment by rpgoldman** [comment]
Actually, we *do* pull from path: in both `find-dot` and `find-neato` the first call is to `find-in-path`, and only if that fails do we search for the executables in likely locations.

I can't speak to any of the design decisions.  It's possible that this was intended to cater to people who were not using graphviz before they installed `cl-dot`, so that graphviz might be installed, but not in their paths. But that's just a guess on my part.  I agree that generally we should find these in the user's path.

There's further confusion, because there were two different bits of code that were used to find the executables: one at initialization and one ... I don't exactly know when. It's all quite confusing.

**Comment by rpgoldman** [comment]
> Exactly.
> 
> The question is: what was that original reason?
> 
> It might be worthwhile to figure that out before gluing more wallpaper over it.

I don't think we will get such an answer: I don't believe @michaelw is reading this repo any more.

One possibility: strip `find-neato` and `find-dot` down to check only the user's `PATH` and the environment variables and drop the bit of code that grovels around likely directories.  That would be fine with me.

Also: Not sure that checking these eagerly at load time makes sense. Why not just check if/when the `dot-graph` function is called?

**Comment by rpgoldman** [comment]
See most recent push for code that only looks for executables as needed.

Have not yet removed the "guess-location" code.  Leave it in case removing it breaks cl-dot for someone?  Remove it in the interests of simplicity?

And ... can anyone test this on Windows? I can't.

**Comment by rpgoldman** [comment]
@mdbergmann @hraban Want to have one more look at this? I think it's ready to go and I will plan to merge it tomorrow.

**Comment by hraban** [comment]
> @mdbergmann @hraban Want to have one more look at this? I think it's ready to go and I will plan to merge it tomorrow.

works 👍 https://github.com/hraban/cl-nix-lite/actions/runs/7981894476
